### PR TITLE
DT-1644 Sync court data between Court Register and Probation

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/resource/SyncResource.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+@Validated
+@RequestMapping(name = "Sync To Delius", path = ["/sync"], produces = [MediaType.APPLICATION_JSON_VALUE])
+class SyncResource(
+  private val courtRegisterSyncService: CourtRegisterSyncService
+) {
+  @PreAuthorize("hasRole('ROLE_MAINTAIN_REF_DATA') and hasAuthority('SCOPE_write')")
+  @Operation(
+    summary = "Sync all court details from HMPPS Court Register",
+    description = "Updates court information, role required is MAINTAIN_REF_DATA",
+    security = [SecurityRequirement(name = "MAINTAIN_REF_DATA", scopes = ["write"])],
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/resource/SyncResource.kt:28]: Ooh I like this, seems like the right place to mention the role and scope.  Should the role be `ROLE_MAINTAIN_REF_DATA` though as that's what is added to the client credentials?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/services/CourtRegisterService.kt [hunk 2]
```diff
@@ -11,6 +12,8 @@ import reactor.core.publisher.Mono
 @Service
 class CourtRegisterService(@Qualifier("courtRegisterApiWebClient") private val webClient: WebClient) {
 
+  private val courts = object : ParameterizedTypeReference<List<CourtDto>>() {}
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/services/CourtRegisterService.kt:15]: Very minor, but maybe name this `courtsType` as `courts` sounds like a `List<CourtDto>>`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/services/CourtRegisterSyncService.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  private fun syncAllCourts(
+    probationCourts: List<CourtFromProbationSystem>,
+    courtRegisterCourts: List<CourtDto>
+  ): SyncStatistics {
+
+    // get all active / inactive courts from Probation (DELIUS)
+    val allCourtsHeldInProbation = probationCourts.map { ProbationService.CourtDataToSync(it) }.associateBy { it.code }
+
+    // get all the courts from the register
+    val allRegisteredCourts: MutableList<ProbationService.CourtDataToSync> = mutableListOf()
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/services/CourtRegisterSyncService.kt:26]: Can you use a map like  you did above for `allCourtsHeldInProbation`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/resource/SyncResource.kt (hunk 1 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/resource/SyncResource.kt:28]: I think Spring actually allows for the missing ROLE at the start (but I do usually add this in as I prefer it as it's more explicit - I will update it).  Also looks like some missing tests on that endpoint.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/services/CourtRegisterService.kt (hunk 2 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/services/CourtRegisterService.kt:15]: Will update

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/services/CourtRegisterSyncService.kt [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    courtRegisterCourts.forEach {
+      allRegisteredCourts.add(ProbationService.CourtDataToSync(it))
+    }
+    val courtRegisterMap = allRegisteredCourts.associateBy { it.code }
+
+    val stats = SyncStatistics()
+    // matches
+    courtRegisterMap.filter { c -> allCourtsHeldInProbation[c.key] != null }
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstodeliusupdate/services/CourtRegisterSyncService.kt:34]: You could maybe move each of these filters up into one line functions with nice names like `matches`, `new`, `missing`?  Just one of those cases where you could do the "remove the comments with self-documenting kind" type approach.

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**